src/shell/shell_cmds.cpp

```diff
@@ -1251,6 +1251,189 @@ void DOS_Shell::CMD_COPY(char * args) {
+                             attributes attribs)
+{
+	char path[DOS_PATHLENGTH + 4], full[DOS_PATHLENGTH];
+	if (!DOS_Canonicalize(args, full) || strrchr(full, '\\') == NULL) {
+		shell->WriteOut(MSG_Get("SHELL_ILLEGAL_PATH"));
+		return false;
+	}
+	bool result = DOS_FindFirst(args, 0xffff & ~DOS_ATTR_VOLUME);
+	if (!result && !optS)
+		return false;
+	char *end = strrchr(full, '\\');
+	if (end == NULL)
+		return false;
+	end++;
+	*end = 0;
+	strcpy(path, full);
+	char name[DOS_NAMELENGTH_ASCII];
+	uint32_t size;
+	uint16_t time, date, fattr;
+	uint8_t attr;
+	bool found = false;
+	while (result) {
+		dta.GetResult(name, size, date, time, attr);
+		if (!((!strcmp(name, ".") || !strcmp(name, "..") ||
+		       strchr(args, '*') != NULL || strchr(args, '?') != NULL) &&
+		      attr & DOS_ATTR_DIRECTORY)) {
+			found = true;
+			strcpy(end, name);
+			if (*full && DOS_GetFileAttr(full, &fattr)) {
+				if (attribs.add_a || attribs.add_s || attribs.add_h || attribs.add_r ||
+				    attribs.min_a || attribs.min_s || attribs.min_h || attribs.min_r) {
+					fattr |= (attribs.add_a ? DOS_ATTR_ARCHIVE : 0);
+					fattr |= (attribs.add_s ? DOS_ATTR_SYSTEM : 0);
+					fattr |= (attribs.add_h ? DOS_ATTR_HIDDEN : 0);
+					fattr |= (attribs.add_r ? DOS_ATTR_READ_ONLY : 0);
+					fattr &= (attribs.min_a ? ~DOS_ATTR_ARCHIVE : 0xffff);
+					fattr &= (attribs.min_s ? ~DOS_ATTR_SYSTEM : 0xffff);
+					fattr &= (attribs.min_h ? ~DOS_ATTR_HIDDEN : 0xffff);
+					fattr &= (attribs.min_r ? ~DOS_ATTR_READ_ONLY : 0xffff);
+					if (DOS_SetFileAttr(full, fattr) && DOS_GetFileAttr(full, &fattr))
+						show_attributes(shell, fattr, full);
+					else
+						shell->WriteOut(MSG_Get("SHELL_CMD_ATTRIB_SET_ERROR"),
+						                full);
+				} else {
+					show_attributes(shell, fattr, full);
+				}
+			} else
+				shell->WriteOut(MSG_Get("SHELL_CMD_ATTRIB_GET_ERROR"),
+				                full);
+		}
+		result = DOS_FindNext();
+	}
+	if (optS) {
+		size_t len = strlen(path);
+		strcat(path, "*.*");
+		bool ret = DOS_FindFirst(path, 0xffff & ~DOS_ATTR_VOLUME);
+		*(path + len) = 0;
+		if (ret) {
+			std::vector<std::string> found_dirs;
+			found_dirs.clear();
+			do { /* File name and extension */
+				DtaResult result;
+				dta.GetResult(result.name, result.size, result.date,
+				              result.time, result.attr);
+
+				if ((result.attr & DOS_ATTR_DIRECTORY) &&
+				    strcmp(result.name, ".") &&
+				    strcmp(result.name, "..")) {
+					strcat(path, result.name);
+					strcat(path, "\\");
+					char *fname = strrchr(args, '\\');
+					if (fname != NULL)
+						fname++;
+					else {
+						fname = strrchr(args, ':');
+						if (fname != NULL)
+							fname++;
+						else
+							fname = args;
+					}
+					strcat(path, fname);
+					found_dirs.push_back(path);
+					*(path + len) = 0;
+				}
+			} while (DOS_FindNext());
+			all_dirs.insert(all_dirs.begin() + 1,
+			                found_dirs.begin(), found_dirs.end());
```

Comment:
This gets pretty nested and hard to read. 
Haven't looked hard enough at how to improve it though.

Reply:
I refactored the nested code somehow which is likely easier to read.